tools/erc-repository-indexer/erc-contract-indexer/src/index.ts

```diff
@@ -0,0 +1,81 @@
+/*-
+ *
+ * Hedera Smart Contracts
+ *
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+import dotenv from 'dotenv';
+import { ByteCodeAnalyzer } from './services/byteCodeAnalyzer';
+import { ConfigService } from './services/config';
+import { ContractScannerService } from './services/contractScanner';
+import { RegistryGenerator } from './services/registryGenerator';
+
+dotenv.config();
+
+export const main = async () => {
+  const configService = new ConfigService();
+  const registryGenerator = new RegistryGenerator();
+  const contractScannerService = new ContractScannerService(
+    configService.getMirrorNodeUrl()
+  );
+  const byteCodeAnalyzer = new ByteCodeAnalyzer();
+
+  try {
+    let next = await configService.resolveStartingPoint(registryGenerator);
+    await processContracts(
+      next,
+      contractScannerService,
+      byteCodeAnalyzer,
+      registryGenerator
+    );
+  } catch (error) {
+    console.error('Error during the indexing process:', error);
+  }
+};
+
+const processContracts = async (
+  next: string | null,
+  contractScannerService: ContractScannerService,
+  byteCodeAnalyzer: ByteCodeAnalyzer,
+  registryGenerator: RegistryGenerator
+) => {
+  do {
+    const fetchContractsResponse =
+      await contractScannerService.fetchContracts(next);
+
+    if (!fetchContractsResponse || !fetchContractsResponse.contracts.length) {
+      console.warn('No contracts found.');
+      return;
+    }
+
+    next = fetchContractsResponse.links.next;
+
+    const ercContracts = await byteCodeAnalyzer.categorizeERCContracts(
+      contractScannerService,
+      fetchContractsResponse.contracts
+    );
+
+    // let the registry update process to run asynchronously in the background
+    registryGenerator.generateErcRegistry(
+      ercContracts.erc20Contracts,
+      ercContracts.erc721Contracts
+    );
+    registryGenerator.updateNextPointer(next);
```

Comment:
Q: should we have a configurable wait time in between batches e.g 1 sec?
This will allow devs to specify how quick they want to run and might help manage rate limiting 🤔

Reply:
Right nice suggestion! Happy to integrate such feature but I wonder though why would someone want to wait between batches? Sure it's a nice feature for users to have the option, but I’m not sure if anyone would prefer the tool to operate more slowly. 

Regarding managing rate limiting, the tool can auto detect rate limiting errors, then it'd pause for a specified duration, and resend the requests independently.

Reply:
Okay, if we've already got a plan for rate limiting support then no need to try to reduce the likelihood of occurrence